compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2082,7 +2082,12 @@ fn add_order_independent_options(
         // sections to ensure we have all the data for PGO.
         let keep_metadata =
             crate_type == CrateType::Dylib || sess.opts.cg.profile_generate.enabled();
-        cmd.gc_sections(keep_metadata);
+
+        if !(crate_type == CrateType::Executable
+            && sess.opts.unstable_opts.export_executable_symbols)
+        {
+            cmd.gc_sections(keep_metadata);
+        }
```

Comment:
Maybe move the `no_gc_sections` call here if `export_executable_symbols` is true.